core/shared/src/main/scala/magnolia.scala

```diff
@@ -362,16 +362,17 @@ object Magnolia {
             )"""
           }
         } else {
-          val defaults = headParamList.fold(List[Tree](q"$scalaPkg.None")) { params =>
-            val x = TermName(c.freshName("x"))
-            val A = TypeName(c.freshName("A"))
-            val dummyArgs = params.filterNot(_.isParamWithDefault).map(p => q"${p.name} = $x")
-            // `A <: Nothing` to avoid dead code warnings
-            val dummy = c.typecheck(q"def $x[$A <: Nothing]($x: $A) = new $genericType(..$dummyArgs)")
-            val defaultParams = dummy.collect { case sel: Select if sel.symbol.isSynthetic => sel }.iterator
-            for (p <- params) yield {
-              if (!p.isParamWithDefault) q"$scalaPkg.None"
-              else q"$scalaPkg.Some(${defaultParams.next()})"
+          val defaults = headParamList.fold[List[Tree]](Nil) { params =>
+            if (params.exists(_.isParamWithDefault)) {
+              val x = TermName(c.freshName("x"))
+              val A = TypeName(c.freshName("A"))
+              val dummyArgs = params.filterNot(_.isParamWithDefault).map(p => q"${p.name} = $x")
+              // `A <: Nothing` to avoid dead code warnings
+              val dummy = c.typecheck(q"def $x[$A <: $scalaPkg.Nothing]($x: $A) = new $genericType(..$dummyArgs)")
+              val defaults = dummy.collect { case sel @ Select(_, name) if name.toString.contains("$default$") => sel }.iterator
```

Comment:
If you wonder why we can't just get the parameters in a list, there is this:
```scala
def x$macro$258[A$macro$259 <: scala.Nothing](x$macro$258: A$macro$259) = {
  val x$2: A$macro$259 = x$macro$258;
  val x$3: A$macro$259 = x$macro$258;
  val x$4: Int = magnolia.tests.Item.`<init>$default$2`;
  new magnolia.tests.Item(x$2, x$4, x$3)
}
```
The compiler rewrote it to preserve execution order when there is a default parameter in the middle.

Reply:
That's crazy. But we're playing with fire, anyway.